pkg/oauthflow/pkce.go

```diff
@@ -18,38 +18,57 @@ package oauthflow
 import (
 	"crypto/sha256"
 	"encoding/base64"
-	"errors"
+	"fmt"
 
+	"github.com/coreos/go-oidc/v3/oidc"
 	"golang.org/x/oauth2"
-)
 
-type PKCEMethod string
+	"github.com/sigstore/fulcio/pkg/log"
+)
 
 const (
-	PKCEPlain PKCEMethod = "plain"
-	PKCES256  PKCEMethod = "S256"
+	PKCEPlain = "plain"
+	PKCES256  = "S256"
 )
 
 type PKCE struct {
 	Challenge string
-	Method    PKCEMethod
+	Method    string
 	Value     string
 }
 
-func NewPKCE(method PKCEMethod) (*PKCE, error) {
-	switch method {
-	case PKCEPlain, PKCES256:
-	default:
-		return nil, errors.New("invalid PKCE method requested")
+func NewPKCE(provider *oidc.Provider) (*PKCE, error) {
+	var providerClaims struct {
+		CodeChallengeMethodsSupported []string `json:"code_challenge_methods_supported"`
 	}
 
-	value, err := randStr()
-	if err != nil {
+	if err := provider.Claims(&providerClaims); err != nil {
+		// will only error out if the JSON was malformed, which shouldn't happen at this point
 		return nil, err
 	}
 
+	var chosenMethod string
```

Comment:
This section terrifies me for downgrade attacks: https://tools.ietf.org/html/rfc7636#section-7.2

Who doesn't support S256? Do we care?